lib/entice/logic/casting.ex

```diff
@@ -8,51 +8,115 @@ defmodule Entice.Logic.Casting do
+  do: Entity.call_behaviour(entity, Casting.Behaviour, {:cast_start, nil, nil, %{target: target, skill: skill}})
 
 
   defmodule Behaviour do
     use Entice.Entity.Behaviour
 
-    def handle_call(
-        {:skill_cast_start, %{target: target, skill: skill}},
-        %Entity{attributes: %{Energy => %Energy{mana: mana}}} = entity) do
-
-      case mana - skill.energy_cost do
-        new_mana when new_mana < 0 -> {:ok, {:error, :not_enough_energy}, entity}
-        new_mana ->
-          case skill.effect_cast_start(entity, target) do
-            {:error, reason} -> {:ok, {:error, reason}, entity}
-            {:ok, %Entity{} = entity, %Entity{} = target} ->
-              target.id |> Entity.attribute_transaction(&Map.merge(&1, target.attributes))
-              {:ok, :ok, entity}
-            result -> raise "Incorrect skill casting result:\nSkill: #{skill.id}\nResult: #{result}"
-          end
+    def init(entity, %Casting{} = casting),
+    do: {:ok, entity |> put_attribute(casting)}
+
+    def terminate(_reason, entity),
+    do: {:ok, entity |> remove_attribute(Casting)}
+
+    def handle_call({:cast_start, cast_callback, recharge_callback, %{target: _target, skill: skill}}, %Entity{attributes: %{
+      Casting => %Casting{recharge_timers: _recharge_timers},
+      Energy => %Energy{mana: mana}}} = entity) do
+
+      response = can_cast?(skill, entity)
+      entity = case response do
+        {:ok, _} ->
+          entity
+          |> reduce_mana(mana - skill.energy_cost)
+          #Start the casting_timer
+          cast_start(skill.cast_time, skill, cast_callback, recharge_callback)
+          entity
+        _ -> entity
+      end
+
+      {:ok, response, entity}
+    end
+
+    @doc "This event triggers when the cast ends, it resets the casting timer, calls the skill's callback, and triggers recharge_end after a while."
+    def handle_event({:cast_end, skill, cast_callback, recharge_callback}, entity) do
+      cast_callback.(skill)
+      recharge_start(skill.recharge_time, skill, recharge_callback)
+      recharge_timers = Map.update(entity.recharge_timers, skill, skill.recharge_time)
+      {:ok, entity |> update_attribute(Casting, fn s -> %Casting{s | casting_timer: nil, recharge_timers: recharge_timers} end)}
+    end
+
+    @doc "This event triggers when a skill's recharge period ends, it resets the recharge timer for the skill."
+    def handle_event({:recharge_end, skill, _recharge_callback}, entity) do
+      recharge_timers = Map.remove(entity.recharge_timers, skill)
+      {:ok, entity |> update_attribute(Casting, fn s -> %Casting{s | recharge_timers: recharge_timers} end)}
+    end
+
+    def handle_event({:after_cast_end, _skill, _recharge_callback}, entity) do
+      {:ok, entity |> update_attribute(Casting, fn s -> %Casting{s | after_cast_timer: nil} end)}
+    end
+
+    defp reduce_mana(entity, new_mana) do
+      Entity.update_attribute(entity, Energy, fn e -> %Energy{e | mana: new_mana} end)
```

Comment:
please don't call the Entity.[something] api from within a behaviour since it's an additional indirection when we can simply update the attributes from within this